### PR TITLE
feat: add a built-in console task

### LINKED ISSUE #5575: Console task
Start a node console with the `HardhatRuntimeEnvironment` initialized and attached to the `global` scope:

```tsx
hardhat console
> console.log(hre.config.solidity)
// "0.8.24"
```

For inspiration consider the HHv2 version of console:

https://github.com/NomicFoundation/hardhat/blob/83721d9cf23fb8d40eb372fea793377eba89e582/packages/hardhat-core/src/builtin-tasks/console.ts#L13

The v3 version of the `run` task is likely a good comparison for integration testing.

We should consider accepting a --config parameter to allow specifying a path to an alternative `hardhat.config.{js,ts}` file.

### PR DESCRIPTION
**Pull request by galargh**
- [ ] Because this PR includes a **bug fix**, relevant tests have been included.
- [x] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.
- [x] I didn't do anything of this.

---

Resolves #5575

This PR adds a `console` task to the next version of `hardhat`. Here's how it looks like:

```
$ # Show console task's help
$ 
$ pnpm hardhat console --help
Opens a hardhat console

Usage: hardhat [GLOBAL OPTIONS] console [--history <STRING>] [--no-compile] [--] [commands]

OPTIONS:

  --history         Path to a history file
  --no-compile      Don't compile before running this task

POSITIONAL ARGUMENTS:

  commands          Commands to run in the console

For global options help run: hardhat --help
$ 
$ # Run the console task and preview the `hre` context variable
$ 
$ pnpm hardhat console
> .help
.break    Sometimes you get stuck, this gets you out
.clear    Break, and also clear the local context
.editor   Enter editor mode
.exit     Exit the REPL
.help     Print this help message
.load     Load JS from a file into the REPL session
.save     Save all evaluated commands in this REPL session to a file

Press Ctrl+C to abort current expression, Ctrl+D to exit the REPL
> hre
HardhatRuntimeEnvironmentImplementation { userConfig: { tasks: [Array], plugins: [Array], privateKey: [Object] }, config: { plugins: [Array], tasks: [Array], path...
```

The task includes the following options:
- `history` - allows specifying the path where REPL history file should be stored. If it's relative, it will be resolved in the context of global hardhat cache dir. If it is explicitly set to an empty string, the history will not be preserved. It defaults to `console-history.txt` in the global hardhat cache dir. This is very helpful during testing, but I thought it might be useful to the users, too.
- `noCompile` - disables compilation task. This is not implemented yet.
- `options` - This is a hidden argument. It is not exposed to the users. It is only used during testing to set custom REPL options. In particular, we override the IO streams during testing because otherwise the test runner hangs indefinitely.
- `commands` - This is a variadic argument. One can provide a list of commands that should be executed in the REPL after the startup. It defaults to running the `.help` command after startup. This is very helpful during testing because we want to run the commands after the entire REPL setup fully complete there, but I thought it might be useful to the users, too. 

#### Testing
- [x] Added tests for the newly created task action; they are passing in CI
- [x] Manually run the newly added task in the `example-project`; `pnpm hardhat console`

### CODE REVIEW

File: v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+const consoleAction: NewTaskActionFunction<ConsoleActionArguments> = async (
+  { commands },
+  _hre,
+) => {
+  const globalCacheDir = await getCacheDir();
+
+  return new Promise<REPLServer>((resolve) => {
+    // Start a new REPL server with the default options
+    const replServer = repl.start();
+
```
Comment (alcuadrado) [inline v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:25]: I think what's missing here is exporting the `hre`. I believe that's done with `replServer.context.hre = hre`.

Comment (alcuadrado) [reply v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:25]: We should also consider exposing each `hre` field in the context, similar to what we do in `v-next/hardhat/src/index.ts`

File: v-next/hardhat/src/internal/builtin-plugins/console/index.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import type { HardhatPlugin } from "@ignored/hardhat-vnext-core/types/plugins";
+
+import { task } from "@ignored/hardhat-vnext-core/config";
+
+const hardhatPlugin: HardhatPlugin = {
+  id: "console",
+  tasks: [
+    task("console", "Opens a hardhat console")
+      .setAction(import.meta.resolve("./task-action.js"))
```
Comment (alcuadrado) [inline v-next/hardhat/src/internal/builtin-plugins/console/index.ts:9]: In v2 we compile the project before running by default, and have a `--no-compile` flag. We should keep that functionality.

File: v-next/hardhat/src/internal/builtin-plugins/console/index.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+      .addVariadicArgument({
```
Comment (alcuadrado) [inline v-next/hardhat/src/internal/builtin-plugins/console/index.ts:10]: I'm not sure what this'd do

**Comment by alcuadrado** [comment]
I did an initial review despite this being a draft. Looking good! I just left some minor comments.

File: v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    // Start a new REPL server with the default options
+    const replServer = repl.start(options);
+
+    // Resolve the task action promise only when the REPL server exits
+    replServer.on("exit", () => {
+      resolve(replServer);
+    });
+
+    // Add the Hardhat Runtime Environment to the REPL context
+    replServer.context.hre = hre;
```
Comment (galargh) [inline v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:48]: @alcuadrado I'm continuing [the dicsussion about the `hre` in the context](https://github.com/NomicFoundation/hardhat/pull/5592#discussion_r1712244971) here since I now added it.

> I think what's missing here is exporting the hre. I believe that's done with replServer.context.hre = here.
> 
> We should also consider exposing each hre field in the context, similar to what we do in v-next/hardhat/src/index.ts

I didn't expose it initially because I assumed we wanted the users to import it explicitly with `const { default: hre } = await import("@ignored/hardhat-vnext");` for example.

Here's also a suggestion for exposing other hre fields in the context. Let me know if we should proceed with it.   

```suggestion
    // Add the Hardhat Runtime Environment to the REPL context
    replServer.context.hre = hre;
    replServer.context.config = config;
    replServer.context.tasks = tasks;
    replServer.context.globalOptions = globalOptions;
    replServer.context.hooks = hooks;
    replServer.context.interruptions = interruptions;
```

File: v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+    afterEach(async function () {
+      // We try to remove the temporary cache dir after each test, but we don't
+      // fail the test if it fails. For example, we have observed that in GHA
+      // on Windows, the temp dir cannot be removed due to permission issues.
+      try {
+        await remove(cacheDir);
+      } catch (error) {
+        log("Failed to remove temporary cache dir", error);
+      }
```
Comment (galargh) [inline v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts:165]: I added a try-catch and a comment about an issue I observed in https://github.com/NomicFoundation/hardhat/actions/runs/10332073748/job/28602920453, for example. Have you seen something similar before? Do you think it's worth ensuring we can always delete the temp dirs in CI? In this particular case I think it might have something to do that the temp dir is on `C:` while the runner is set up to have perms and execute in the context of drive `D:`.

File: v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  // If noCompile is false, run the compile task first
+  if (!noCompile) {
+    // todo: run compile task
+  }
+
+  return new Promise<REPLServer>(async (resolve) => {
+    // Start a new REPL server with the default options
+    const replServer = repl.start(options);
```
Comment (galargh) [inline v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:40]: Unfortunately, the only option to create an instance of REPL is to start it immediately. This is because the constructor of the REPL class is private. Ideally, we would only start it after the rest of the setup is complete.

**Review by galargh — COMMENTED** [review]
Thanks for the initial review! I usually tend to use drafts for work-in-progress proposals which is pending my self-review. Here, it turned out not to be too far off from the final "ready for review" state ;)

File: v-next/hardhat/src/internal/builtin-plugins/console/index.ts (hunk 2 above)
Comment (alcuadrado) [reply v-next/hardhat/src/internal/builtin-plugins/console/index.ts:9]: correct

File: v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+  });
+
+  describe("history", function () {
+    let cacheDir: string;
+    let history: string;
+
+    beforeEach(async function () {
+      cacheDir = await fsPromises.mkdtemp(
+        path.resolve(os.tmpdir(), "console-action-test-"),
+      );
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts:153]: Yes, let's add it to the utils instead. Our intention is to avoid using `fs` anywhere except that one module. We'll soon have a linter rule for that.

But let's do it in a follow up PR. Can you do it/create an issue about it?

File: v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts (hunk 4 above)
Comment (alcuadrado) [reply v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:48]: > I didn't expose it initially because I assumed we wanted the users to import it explicitly with const { default: hre } = await import("@ignored/hardhat-vnext"); for example.

We want that for tests/scripts, but I think in the console it makes sense to automatically import them, similarly to what node's own console does.

Comment (alcuadrado) [reply v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:48]: Let's commit that change, yes

File: v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (alcuadrado) [reply v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts:153]: @schaable tagging you just for visibility.

File: v-next/hardhat/src/internal/builtin-plugins/console/index.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+const hardhatPlugin: HardhatPlugin = {
+  id: "console",
+  tasks: [
+    task("console", "Opens a hardhat console")
+      .setAction(import.meta.resolve("./task-action.js"))
+      .addOption({
+        name: "history",
+        description: "Path to a history file",
+        defaultValue: "console-history.txt",
+      })
```
[the remark this answers is not in the document]
Comment (alcuadrado) [reply v-next/hardhat/src/internal/builtin-plugins/console/index.ts:14]: Let's keep them. But as a small note, it can be faster to sync on these things on slack first.

File: v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import debug from "debug";
+
+const log = debug("hardhat:core:tasks:console");
+
+interface ConsoleActionArguments {
+  commands: string[];
+  history: string;
+  noCompile: boolean;
+  // We accept ReplOptions as an argument to allow tests overriding the IO streams
+  options?: repl.ReplOptions;
```
Comment (alcuadrado) [inline v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:17]: This is really interesting! I never thought about it.

File: v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+const consoleAction: NewTaskActionFunction<ConsoleActionArguments> = async (
+  { commands, history, noCompile, options },
+  hre,
+) => {
+  // Resolve the history path if it is not empty
+  let historyPath: string | undefined;
+  if (history !== "") {
+    const globalCacheDir = await getCacheDir();
```
Comment (alcuadrado) [inline v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:27]: Given that we have a local cache dir per project, I think it's better if we use that one.

If it's not available yet, let's just create an issue to make sure we get back to this.

File: v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    historyPath = path.isAbsolute(history)
+      ? history
+      : path.resolve(globalCacheDir, history);
+  }
+
+  // If noCompile is false, run the compile task first
+  if (!noCompile) {
+    // todo: run compile task
```
Comment (alcuadrado) [inline v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:35]: Let's create an issue so that we don't forget about it. Do you mind doing it?

File: v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts (hunk 6 above)
Comment (alcuadrado) [reply v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:40]: It's crazy, but unfortunately, not surprising.

File: v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts (hunk 5 above)
Comment (alcuadrado) [reply v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts:165]: that's actually sirprising, but I think we can live with it.

Maybe we should move it to the test utils package so that we don't hit this edge case again.

**Review by alcuadrado — APPROVED** [review]
This looks good! :)

I left a few comments, mostly about things that can be done later. Do you mind creating issues so that we don't loose track of them?

**Review by schaable — COMMENTED** [review]
Nothing to add to Pato's comments. This looks great, nice work!

File: v-next/hardhat/src/internal/builtin-plugins/console/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (galargh) [reply v-next/hardhat/src/internal/builtin-plugins/console/index.ts:14]: Got it! Thanks :)

File: v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts (hunk 10 above)
Comment (galargh) [reply v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:27]: Of course, the issue created at https://github.com/NomicFoundation/hardhat/issues/5599. I'll also add an inline comment about it.

File: v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts (hunk 11 above)
Comment (galargh) [reply v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts:35]: Not at all, the issue's here - https://github.com/NomicFoundation/hardhat/issues/5600

File: v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (galargh) [reply v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts:154]: I created the issue here - https://github.com/NomicFoundation/hardhat/issues/5601

File: v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts (hunk 5 above)
Comment (galargh) [reply v-next/hardhat/test/internal/builtin-plugins/console/task-action.ts:166]: I created an issue for this - https://github.com/NomicFoundation/hardhat/issues/5602 If I cannot resolve it, I'm going to add the utility function for this.
